### DISCUSSION TITLE
Would disabling URL rules improve performance?

### DISCUSSION THREAD

**Discussion by psychosispy** [uBlockOrigin/uBlock-issues#3182] [Ideas] [answered]
I've implemented more comprehensive URL blocking rules on my router compared to UBO. If I keep UBO's element rules unchanged but disable UBO's URL rules and let the upstream router handle all URL filtering, would there be any difference in blocking effectiveness? Is it possible to improve performance and reduce power consumption on mobile devices?

If this idea is feasible, could an option to ignore URL filtering rules be added? Thank you.

**Comment by Yuki2718** [comment, marked as answer]
This is often heard flawed idea. Assume you block ALL ad-, tracking-, annoyances- , and malicious servers on your router<sup>1</sup>. Every time your browser or OS first has to query DNS record of them unless already cached and only then the requests will be sink-holed. No query ever happens if blocked on browser. People should ditch the naive idea as if uBO matches each and every requests against every rules. Do you see delay in Google search as the number of websites is growing? uBO's matching is the same, in a nutshell, uBO discards rules not having chance to match a request in early stages and thus most of rules are as if non-existent. Coming Chromium MV3 visualizes (sort of) this - all the domain rules will be compressed and counted as 1 rule.
___
<sub>1: This is actually never possible. Even if you added all the lists on earch, it's far from the perfect - this is why we utilize various regex and generic rules to catch those servers and each of such a rule is equivalent to thousands of domain rules. And not all ads and trackers can be blocked in domain-level resolution. For example, FingerprintJS is usually implemented in first-party script in the site you visit.</sub>

**Reply by psychosispy** [reply]
Thank you for your answer, which helped me dispel some unrealistic ideas. I also wanted to ask you, should the excessive use of generic element matching rules be different from URLs? Would it affect the overall performance of all websites?

**Reply by Yuki2718** [reply]
Depends on the type of generic cosmetic filters. Low-generic cosmetic filters such as `##.ad` don't affect as they will be injected only if matches are found, but highly generic ones [can affect](https://github.com/uBlockOrigin/uBlock-issues/discussions/1998) if the number is too large.

**Reply by psychosispy** [reply]
Thank you for your answer

**Comment by gorhill** [comment]
> more comprehensive URL blocking rules on my router compared to UBO

How is it more comprehensive?

**Reply by psychosispy** [reply]
Sorry, my description might not be very accurate (using translation). The rules on the router include various types of terminals and some supplements. The rules used by UBO are targeted at mobile or PC terminals, but in reality, The router can only block individual domains and cannot perform advanced operations

**Reply by gorhill** [reply]
Ok so not "URL blocking rules", it's just list of domain names. A reference answer comparing uBO's filtering capabilities to that of DNS-based blocking is <https://github.com/gorhill/uBlock/wiki/About-%22Why-uBlock-Origin-works-so-much-better-than-Pi%E2%80%91hole-does%3F%22>.

**Reply by Yuki2718** [reply]
Maybe regex can be added, though is a form of Pattern-based filtering.

**Reply by psychosispy** [reply]
The rules of uBO include many basic rules like ||a.com^, which overlap with router DNS rules. Initially, I wanted to ask whether removing these overlaps would improve performance, but my description was poor. I'm sorry
